frameworks/Swift/swift-nio/Sources/swift-nio-tfb-default/main.swift

```diff
@@ -0,0 +1,114 @@
+
+import Foundation
+import NIO
+import NIOHTTP1
+
+struct JSONTestResponse: Encodable {
+	let message = "Hello, World!"
+}
+
+enum Constants {
+	static let httpVersion = HTTPVersion(major: 1, minor: 1)
+	static let serverName = "SwiftNIO"
+
+	static let plainTextResponse: StaticString = "Hello, World!"
+	static let plainTextResponseLength = plainTextResponse.count
+	static let plainTextResponseLengthString = String(plainTextResponseLength)
+
+	static let jsonResponseLength = try! JSONEncoder().encode(JSONTestResponse()).count
+	static let jsonResponseLengthString = String(jsonResponseLength)
+}
+
+private final class HTTPHandler: ChannelInboundHandler {
+
+	public typealias InboundIn = HTTPServerRequestPart
+	public typealias OutboundOut = HTTPServerResponsePart
+
+	let dateFormatter = RFC1123DateFormatter()
+	let jsonEncoder = JSONEncoder()
+
+	func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+		let reqPart = self.unwrapInboundIn(data)
+
+		switch reqPart {
+		case .head(let request):
+			switch request.uri {
+			case "/plaintext":
+				processPlaintext(ctx: ctx)
+			case "/json":
+				processJSON(ctx: ctx)
+			default:
+				_ = ctx.close()
+			}
+		case .body:
+			break
+		case .end:
+			_ = ctx.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+		}
+	}
+
+	private func processPlaintext(ctx: ChannelHandlerContext) {
+		let responseHead = plainTextResponseHead(contentLength: Constants.plainTextResponseLengthString)
+		ctx.write(self.wrapOutboundOut(.head(responseHead)), promise: nil)
+
+		var bodyBuffer = ctx.channel.allocator.buffer(capacity: Constants.plainTextResponseLength)
+		bodyBuffer.write(staticString: Constants.plainTextResponse)
+		ctx.write(self.wrapOutboundOut(.body(.byteBuffer(bodyBuffer))), promise: nil)
+	}
+
+	private func processJSON(ctx: ChannelHandlerContext) {
+		let responseHead = jsonResponseHead(contentLength: Constants.jsonResponseLengthString)
+		ctx.write(self.wrapOutboundOut(.head(responseHead)), promise: nil)
+
+		var bodyBuffer = ctx.channel.allocator.buffer(capacity: Constants.jsonResponseLength)
+		let responseData = try! jsonEncoder.encode(JSONTestResponse())
+		bodyBuffer.write(bytes: responseData)
+		ctx.write(self.wrapOutboundOut(.body(.byteBuffer(bodyBuffer))), promise: nil)
+	}
```

Comment:
Depends on whether you consider this "cheating", but given that both the plaintext and JSON responses are static, they could be computed at `handlerAdded` time and stored in byte buffers on the handler. That will avoid many repeated allocations (in calls to `allocator.buffer`) and also avoid repeated copies.

Reply:
To be clear, I mean this:

```swift
var cachedJSONResponse: ByteBuffer?
var cachedPlaintextResponse: ByteBuffer?

func handlerAdded(ctx: ChannelHandlerContext) {
    self.cachedJSONResponse = ctx.channel.allocator.buffer(capacity: Constants.jsonResponseLength)
    self.cachedJSONResponse.write(bytes: try! jsonEncoder.encode(JSONTestResponse()))
    self.cachedPlaintextResponse = ctx.channel.allocator.buffer(capacity: Constants.plainTextResponseLength)
    self.cachedPlaintextResponse.write(staticString: Constants.plainTextResponse)
}
```

then using those two `self` properties directly.